### PR TITLE
ADM-434 - [report] - [backend] Jira board calculate Velocity

### PR DESCRIPTION
**Pull request by Eason-T**
## Summary
1. add Velocity calculation logic
2. add new FeignClient to get Chinese holiday
...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: backend/src/main/java/heartbeat/service/board/jira/JiraService.java [hunk 1]
```diff
@@ -337,7 +353,181 @@ private List<TargetField> getTargetIssueField(Map<String, IssueField> fields) {
 
 	public Cards getStoryPointsAndCycleTime(StoryPointsAndCycleTimeRequest request,
 			List<RequestJiraBoardColumnSetting> boardColumns, List<String> users) {
-		return Cards.builder().build();
+		CardCustomFieldKey cardCustomFieldKey = saveCustomFieldKey(request);
+		BoardType boardType = BoardType.fromValue(request.getType());
+		URI baseUrl = urlGenerator.getUri(request.getSite());
+		BoardRequestParam boardRequestParam = BoardRequestParam.builder()
+			.boardId(request.getBoardId())
+			.email("")
```
Comment (wangwang-zhang) [inline backend/src/main/java/heartbeat/service/board/jira/JiraService.java:361]: why empty

File: backend/src/main/java/heartbeat/client/HolidayFeignClient.java [hunk 2]
```diff
@@ -1,16 +1,14 @@
...
 import org.springframework.web.bind.annotation.PathVariable;
 
-import java.util.List;
-
 @FeignClient(value = "holidayFeignClient", url = "https://raw.githubusercontent.com/NateScarlet/holiday-cn/master")
 public interface HolidayFeignClient {
 
 	@GetMapping(path = "/{year}.json")
-	List<HolidayDTO> getHoliday(@PathVariable String year);
+	HolidaysResponseDTO getHolidays(@PathVariable String year);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/client/HolidayFeignClient.java:12]: Keep `List<HolidayDTO>` for this

File: backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java [hunk 3]
```diff
@@ -68,13 +67,12 @@ public BuildKiteResponse fetchPipelineInfo(PipelineParam pipelineParam) {
...
 				.collect(Collectors.toList());
-			log.info("[BuildKite] Successfully get buildKite pipelineInfo, pipelineInfoList size is:"
-					+ buildKiteInfoList.size());
+			log.info("Successfully get buildKite pipelineInfo, pipelineInfoList size is:" + buildKiteInfoList.size());
 
 			return BuildKiteResponse.builder().pipelineList(buildKiteInfoList).build();
 		}
 		catch (FeignException e) {
-			log.error("[BuildKite] Failed when call BuildKite", e);
+			log.error("Failed when call BuildKite", e);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java:75]: update to `Failed to call Buildkite_pipelineParam: {}, e: {}) pipelineParas.xxx, e.message)`

File: backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java [hunk 4]
```diff
@@ -160,19 +157,19 @@ private CompletableFuture<List<BuildKiteBuildInfo>> getBuildKiteStepsAsync(Strin
...
 				log.info(
-						"[BuildKite] Successfully get paginated pipeline steps info_token:{},orgId: {},pipelineId: {},pipeline steps size: {},page:{}",
+						"Successfully get paginated pipeline steps info_token:{},orgId: {},pipelineId: {},pipeline steps size: {},page:{}",
 						partialToken, organizationId, pipelineId, pipelineStepsInfo.size(), page);
 				return pipelineStepsInfo;
 			}
 			catch (RequestFailedException e) {
 				log.error(
-						"[BuildKite] Failed get build kite steps page_token: {},orgId: {},pipelineId: {}, exception occurred: {},page: {}",
+						"Failed get build kite steps page_token: {},orgId: {},pipelineId: {}, exception occurred: {},page: {}",
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java:172]: Failed to get ....

File: backend/src/test/java/heartbeat/service/report/WorkDayTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+	@Test
+	void shouldReturnRightWhenCalculateWorkDaysBetween() {
+		String year = "2020";
+		List<HolidayDTO> holidayDTOList = List.of(
+				HolidayDTO.builder().date("2020-01-01").name("元旦").isOffDay(true).build(),
+				HolidayDTO.builder().date("2020-01-24").name("春节").isOffDay(true).build(),
+				HolidayDTO.builder().date("2020-01-25").name("春节").isOffDay(true).build());
+		long startTime = LocalDate.parse("2020-01-01", DateTimeFormatter.ISO_DATE)
+			.atStartOfDay(ZoneOffset.UTC)
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/report/WorkDayTest.java:63]: duplicated code

File: backend/src/test/java/heartbeat/service/report/WorkDayTest.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+		boolean result1 = workDay.verifyIfThisDayHoliday(holidayTime);
+		boolean result2 = workDay.verifyIfThisDayHoliday(workdayTime);
+
+		Assertions.assertTrue(result1);
+		Assertions.assertFalse(result2);
+	}
+
+	@Test
+	void shouldReturnRightWhenCalculateWorkDaysBetween() {
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/report/WorkDayTest.java:56]: should return right?  what is right?

File: backend/src/test/java/heartbeat/util/BoardUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+		Assertions.assertEquals(statusChangedItemsExpect, result);
+
+	}
+
+	@Test
+	void shouldReturnCardTimeForEachStep() {
+		List<StatusChangedItem> statusChangedItems = StatusChangedArrayItemsFixture.STATUS_CHANGED_ITEMS_LIST();
+		when(workDay.calculateWorkDaysBy24Hours(anyLong(), anyLong())).thenReturn(2.0);
+
+		List<CycleTimeInfo> expect = List.of(CycleTimeInfo.builder().column("UNKNOWN").day(4.0).build(),
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/util/BoardUtilTest.java:44]: `4.0` duplicated

**Review by guzhongren — COMMENTED** [review]
Need some change

File: backend/build.gradle [hunk 8]
```diff
@@ -58,7 +58,8 @@ pmd {
 def testIgnoreList = [
   "**/HeartbeatApplication**",
   '**/config/**',
-  '**/decoder/**'
+  '**/decoder/**',
+  '**/TimeUtil**',
```
[the remark this answers is not in the document]
Comment (Eason-T) [reply backend/build.gradle]: decoder test will add to other PR

File: backend/src/test/java/heartbeat/service/report/WorkDayTest.java (hunk 6 above)
Comment (Eason-T) [reply backend/src/test/java/heartbeat/service/report/WorkDayTest.java:56]: done

File: backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java (hunk 4 above)
Comment (Eason-T) [reply backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java:172]: done

File: backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java (hunk 3 above)
Comment (Eason-T) [reply backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java:75]: done

File: backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java [hunk 9]
```diff
@@ -49,17 +49,16 @@ public class BuildKiteService {
 	public BuildKiteResponse fetchPipelineInfo(PipelineParam pipelineParam) {
 		try {
 			String buildKiteToken = "Bearer " + pipelineParam.getToken();
-			log.info("[BuildKite] Start to query token permissions" + TokenUtil.mask(pipelineParam.getToken()));
+			log.info("Start to query token permissions" + TokenUtil.mask(pipelineParam.getToken()));
```
[the remark this answers is not in the document]
Comment (Eason-T) [reply backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java:52]: done

File: backend/src/test/java/heartbeat/service/report/WorkDayTest.java (hunk 5 above)
Comment (Eason-T) [reply backend/src/test/java/heartbeat/service/report/WorkDayTest.java:63]: done

File: backend/src/test/java/heartbeat/util/BoardUtilTest.java (hunk 7 above)
Comment (Eason-T) [reply backend/src/test/java/heartbeat/util/BoardUtilTest.java:44]: done

File: frontend/vite.config.ts [hunk 10]
```diff
@@ -6,7 +6,7 @@ export default defineConfig({
   server: {
     port: 4321,
     proxy: {
-      '/api/v1': 'http://localhost:4322',
+      '/api/v1': 'http://13.215.41.120:4323',
```
Comment (guzhongren) [inline frontend/vite.config.ts:9]: Revert this IP in your commit history

**Review by guzhongren — COMMENTED** [review]
1. Revert IP address in the commit history
2. Write the calculation doc for your cards

**Review by wangwang-zhang — APPROVED** [review]
LGTM

**Pull request merged**